rclpy/src/rclpy/_rclpy.c

```diff
@@ -3303,34 +3305,35 @@ rclpy_take_raw_with_info(rcl_subscription_t * subscription, rmw_message_info_t *
       RCLError,
       "Failed to initialize message: %s", rcl_get_error_string().str);
     rcl_reset_error();
-    rmw_ret_t r_fini = rmw_serialized_message_fini(&msg);
-    if (r_fini != RMW_RET_OK) {
-      PyErr_Format(RCLError, "Failed to deallocate message buffer: %d", r_fini);
-    }
-    return NULL;
+    goto cleanup;
   }
 
   ret = rcl_take_serialized_message(subscription, &msg, message_info, NULL);
   if (ret != RCL_RET_OK) {
-    PyErr_Format(
-      RCLError,
-      "Failed to take_serialized from a subscription: %s", rcl_get_error_string().str);
-    rcl_reset_error();
-    rmw_ret_t r_fini = rmw_serialized_message_fini(&msg);
-    if (r_fini != RMW_RET_OK) {
-      PyErr_Format(RCLError, "Failed to deallocate message buffer: %d", r_fini);
+    if (ret == RCL_RET_SUBSCRIPTION_TAKE_FAILED) {
+      python_bytes = Py_None;
+      Py_INCREF(Py_None);
+    } else {
+      PyErr_Format(
+        RCLError,
+        "Failed to take_serialized from a subscription: %s", rcl_get_error_string().str);
+      rcl_reset_error();
     }
-    return NULL;
+    goto cleanup;
   }
-  PyObject * python_bytes = PyBytes_FromStringAndSize((char *)(msg.buffer), msg.buffer_length);
-  rmw_ret_t r_fini = rmw_serialized_message_fini(&msg);
-  if (r_fini != RMW_RET_OK) {
-    PyErr_Format(RCLError, "Failed to deallocate message buffer: %d", r_fini);
-    if (python_bytes) {
+
+  python_bytes = PyBytes_FromStringAndSize((char *)(msg.buffer), msg.buffer_length);
+
+cleanup:
+  ret = rmw_serialized_message_fini(&msg);
+  if (ret != RMW_RET_OK) {
+    PyErr_Format(RCLError, "Failed to deallocate message buffer: %d", ret);
+    if (python_bytes && Py_None != python_bytes) {
```

Comment:
I actually imitated the logic in your pull request (https://github.com/ros2/rclpy/pull/898) for fixing the serialized memory leak. There it returns the serialized bytes even if message deallocation fails. 

But now, looking at what could make message deallocation fail - seems that it can only fail if memory has become corrupted (https://github.com/ros2/rcutils/blob/foxy/src/allocator.c#L92). And I suppose it would indeed be better to propagate the error in this case, as behavior has become undefined. 

Made it so. 

Question - should it also invoke `rcl_reset_error()`?

Reply:
@suurjaak 

Thank you. To keep the same behavior, I'll update https://github.com/ros2/rclpy/pull/898.

> Question - should it also invoke `rcl_reset_error()`?

I think the answer is Yes.

Reply:
> I think the answer is Yes.

Made it so.